pylearner.ts

```diff
@@ -1990,7 +2183,125 @@ function parseProofOutline(stmts: Statement[], i: number, precededByAssert: bool
+    if(stmt instanceof ExpressionStatement) {
+      if(stmt.expr instanceof AssignmentExpression && stmt.expr.op == '=' && stmt.expr.lhs instanceof VariableExpression) {
+        if(stmt.expr.rhs instanceof VariableExpression) 
+          stmt.mayAliasRelation = stmt.mayAliasRelation.addMayAliases(stmt.expr.lhs.name,stmt.expr.rhs.name);
+        else stmt.mayAliasRelation = stmt.mayAliasRelation.removeVariableInMayAliasSets(stmt.expr.lhs.name);   
+      }
+    }
+    else if(stmt instanceof IfStatement) {
+        if(stmt.thenBody instanceof BlockStatement && stmt.elseBody instanceof BlockStatement &&  stmt.elseBody != null && stmt.thenBody != null) {
+        const stmtsThenBlock = stmt.thenBody.stmts;
+        const stmtsElseBlock = stmt.elseBody.stmts;
+        let mayAliasRelationThenBlock = performAliasAnalysis(stmtsThenBlock,0,stmt.mayAliasRelation)!;
+        let mayAliasRelationElseBlock = performAliasAnalysis(stmtsElseBlock,0,stmt.mayAliasRelation)!;
+        if(mayAliasRelationThenBlock&&mayAliasRelationElseBlock) {
+          const scopeThenBody = stmt.thenBody.scope!
+          const localVars = getLocalVariablesInScope(scopeThenBody);
+          const outerscopeVars = getOuterScopeVariables(scopeThenBody.outerScope!);
+          let varsToKeep = outerscopeVars.subtract(localVars);
+          const mayAliasRelationThenBlockWithVarsToKeep = mayAliasRelationThenBlock.keepVariablesInMayAliasSets(varsToKeep);
+          const mayAliasRelationElseBlockWithVarsToKeep = mayAliasRelationElseBlock.keepVariablesInMayAliasSets(varsToKeep);
+          let combinedMayAliasRelation = mayAliasRelationThenBlockWithVarsToKeep.combineWithMayAliasRelation(mayAliasRelationElseBlockWithVarsToKeep);
+          return performAliasAnalysis(stmts,i+1,combinedMayAliasRelation);
+        } 
+      }
+    }
+    else if(stmt instanceof WhileStatement){
+      if (stmt.body instanceof BlockStatement && stmt.body != null){
+        const stmtsBodyBlock = stmt.body.stmts;
+        let mayAliasRelationWhileBlock = performAliasAnalysis(stmtsBodyBlock,0,stmt.mayAliasRelation);
+        let mayAliasRelationWhileBlockCheck = performAliasAnalysis(stmtsBodyBlock,0,mayAliasRelationWhileBlock);
+        while (mayAliasRelationWhileBlock.mayAliasSets.toJS().toString() != mayAliasRelationWhileBlockCheck.mayAliasSets.toJS().toString()) {
+          mayAliasRelationWhileBlock = mayAliasRelationWhileBlockCheck;
+          mayAliasRelationWhileBlockCheck = performAliasAnalysis(stmtsBodyBlock,0,mayAliasRelationWhileBlock);
+        }
+        if(mayAliasRelationWhileBlock&&mayAliasRelationWhileBlockCheck) {
+          const scope = stmt.body.scope
+          const localVars = getLocalVariablesInScope(scope!);
+          const outerscopeVars = getOuterScopeVariables(scope!.outerScope!);
+          let varsToKeep = outerscopeVars.subtract(localVars); 
+          const mayAliasRelationWithVarsToKeep= mayAliasRelationWhileBlock.keepVariablesInMayAliasSets(varsToKeep);
```

Comment:
Does it not make more sense to remove the block-local variables at each iteration? I suggest you introduce a `performAliasAnalysisOnBlock` function and use it for the body of a `while`, as well as for the branches of an `if`.